### PR TITLE
add-check-for-master-api

### PR DESCRIPTION
**Pull request by calvix**
add an option to enable k8s API check

also, add k8s health check for kubelet

towards https://github.com/giantswarm/giantswarm/issues/4793

### CODE REVIEW

**Comment by calvix** [comment]
```
--- FAIL: Test_Flannel_ParseIP (0.00s)
panic: json: cannot unmarshal string into Go struct field Service.CheckAPI of type bool [recovered]
	panic: json: cannot unmarshal string into Go struct field Service.CheckAPI of type bool

goroutine 22 [running]:
testing.tRunner.func1(0xc0000e6600)
	/usr/local/go/src/testing/testing.go:792 +0x6a7
panic(0xb6dfe0, 0xc0000b6be0)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9
github.com/giantswarm/k8s-kvm-health/vendor/github.com/giantswarm/microkit/flag.Init(0xb2d440, 0xc000093080)
	/go/src/github.com/giantswarm/k8s-kvm-health/vendor/github.com/giantswarm/microkit/flag/flag.go:33 +0x4cb
github.com/giantswarm/k8s-kvm-health/flag.New(0xaee1bd)
	/go/src/github.com/giantswarm/k8s-kvm-health/flag/flag.go:14 +0x94
github.com/giantswarm/k8s-kvm-health/service.Test_Flannel_ParseIP.func1(0xc000168000, 0x61, 0x70, 0xc00014a280, 0x139, 0x140, 0xdf3fd0)
	/go/src/github.com/giantswarm/k8s-kvm-health/service/flannel_test.go:21 +0xc8
github.com/giantswarm/k8s-kvm-health/service.Test_Flannel_ParseIP(0xc0000e6600)
	/go/src/github.com/giantswarm/k8s-kvm-health/service/flannel_test.go:113 +0x54d
testing.tRunner(0xc0000e6600, 0xc24aa8)
	/usr/local/go/src/testing/testing.go:827 +0x163
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:878 +0x651
FAIL	github.com/giantswarm/k8s-kvm-health/service	0.034s
?   	github.com/giantswarm/k8s-kvm-health/service/healthz	[no test files]
?   	github.com/giantswarm/k8s-kvm-health/service/healthz/kvm	[no test files]
```
any idea how could i go around this?
ping @JosephSalisbury

File: service/healthz/healthz.go [hunk 1]
```diff
@@ -30,9 +31,10 @@ func New(config Config) (*Service, error) {
 
 	var kvmService healthz.Service
 	{
-		kvmServiceConfig := kvm.DefaultConfig()
+		var kvmServiceConfig kvm.Config
```
Comment (JosephSalisbury) [inline service/healthz/healthz.go]: I'd expect
```
kvmServiceConfig := kvm.Config{
    IP: config.IPaddress,
    Logger: config.Logger,
    CheckAPI: config.CheckAPI
}

kvmService, err = ...
```
this feels like the old style we had a while back

File: service/healthz/kvm/kvm.go [hunk 2]
```diff
@@ -60,29 +58,46 @@ func New(config Config) (*Service, error) {
...
 	response := healthz.Response{
 		Description: Description,
-		Failed:      failed,
-		Message:     message,
+		Failed:      pingFailed,
+		Message:     pingMsg,
 		Name:        Name,
 	}
 
+	// check kubelet only if ping succeeded
```
Comment (JosephSalisbury) [inline service/healthz/kvm/kvm.go]: Why do we check only if the ping succeeded?

File: service/healthz/kvm/kvm.go [hunk 3]
```diff
@@ -60,29 +58,46 @@ func New(config Config) (*Service, error) {
...
+		response.Failed = apiFailed
+		response.Message += apiMsg
+	}
+
 	return response, nil
 }
 
-// implementation fo the interface healthz logic
-func (s *Service) healthCheck() (bool, string) {
+// implementation fo the interface healthz logic for ping check
```
Comment (tuommaki) [inline service/healthz/kvm/kvm.go]: I'm not fully sure if this comment provides much value here? Maybe delete?

File: service/healthz/kvm/kvm.go [hunk 4]
```diff
@@ -108,3 +123,23 @@ func (s *Service) healthCheck() (bool, string) {
 	// exit
 	return failed, message
 }
+
+// implementation of the interface healthz logic for http  check
```
Comment (tuommaki) [inline service/healthz/kvm/kvm.go]: Regarding this comment, same as above. I'm not sure if this is really useful.

File: service/healthz/kvm/kvm.go [hunk 5]
```diff
@@ -60,29 +58,46 @@ func New(config Config) (*Service, error) {
...
-		logger: config.Logger,
+		checkAPI: config.CheckAPI,
+		ip:       config.IP,
+		logger:   config.Logger,
 	}
 
 	return newService, nil
 }
 
 // GetHealthz implements the health check for network interface.
```
Comment (tuommaki) [inline service/healthz/kvm/kvm.go]:
```suggestion
// GetHealthz Provides Healthz implementation to check health status of network
// interface. It performs following checks in given order:
//  - Ping configured IP. 
//  - Check that Kubelet instance in configured IP responds to HTTP request.
//  - Check that K8s API in configured IP responds to HTTPS request. 
```

File: service/service.go [hunk 6]
```diff
@@ -62,6 +62,11 @@ func New(config Config) (*Service, error) {
 	{
 		healthzConfig := healthz.DefaultConfig()
 
+		if config.Flag.Service.CheckAPI == "true" {
```
Comment (tuommaki) [inline service/service.go]:
```suggestion
		if config.Flag.Service.CheckAPI == strings.ToLower("true") {
```
Having case insensitivity here may save some time in the future when being unsure if it should be `TRUE` or `true`.

File: service/service.go [hunk 7]
```diff
@@ -62,6 +62,11 @@ func New(config Config) (*Service, error) {
 	{
 		healthzConfig := healthz.DefaultConfig()
```
Comment (tuommaki) [inline service/service.go]: While here, maybe also fix this to be proper struct initialization as we have agreed here: https://github.com/giantswarm/fmt/blob/master/go/struct_initialization.md ?

**Review by tuommaki — APPROVED** [review]
Couple nitpicks / suggestions on comments and env variable handling. As Joe already pointed for other section, aligning struct initialization to current practices makes sense here as this change is also otherwise touching this code quite widely. In general this LGTM.

File: service/healthz/kvm/kvm.go (hunk 2 above)
Comment (calvix) [reply service/healthz/kvm/kvm.go]: Why would we waste checking TCP layer 7 when TCP layer 4 is not working?
That's like seeing your car has no wheels, but you wanna go and check if the autonomous system can drive the car.


If the ping fails, the HTTP check will fail too.

**Comment by calvix** [comment]
PTAL

**Pull request merged**